# Creating ecoinvent 3 migrations

This notebook describes how migrations were created for ecoinvent versions 3.1 and 3.2 to 3.3. Migrations were created using the ecoinvent change spreadsheets, and some manual matching. Manual matching occurred outside of this notebook.

Start by setting up a new project:

In [1]:
from brightway2 import *
projects.set_current("ecoinvent updates")

In [2]:
bw2setup()

Creating default biosphere



Writing activities to SQLite3 database:
0%                          100%
[##########                    ] | ETA[sec]: 0.371 

Applying strategy: normalize_units
Applying strategy: drop_unspecified_subcategories
Applied 2 strategies in 0.01 seconds


[##############################] | ETA[sec]: 0.000 
Total time elapsed: 0.426 sec


Title: Writing activities to SQLite3 database:
  Started: 12/06/2016 10:51:41
  Finished: 12/06/2016 10:51:41
  Total time elapsed: 0.426 sec
  CPU %: 101.800000
  Memory %: 0.409055
Created database: biosphere3
Creating default LCIA methods

Applying strategy: normalize_units
Applying strategy: set_biosphere_type
Applying strategy: drop_unspecified_subcategories
Applying strategy: link_iterable_by_fields
Applied 4 strategies in 0.97 seconds
Wrote 718 LCIA methods with 178008 characterization factors
Creating core data migrations



In [3]:
ei = SingleOutputEcospold2Importer(
    "/Users/cmutel/Documents/LCA Documents/Ecoinvent/3.1/cutoff/datasets", 
    "3.1 cutoff"
)
ei.apply_strategies()
ei.statistics()

Extracting XML data from 11301 datasets
Extracted 11301 datasets in 39.37 seconds
Applying strategy: normalize_units
Applying strategy: remove_zero_amount_coproducts
Applying strategy: remove_zero_amount_inputs_with_no_activity
Applying strategy: remove_unnamed_parameters
Applying strategy: es2_assign_only_product_with_amount_as_reference_product
Applying strategy: assign_single_product_as_activity
Applying strategy: create_composite_code
Applying strategy: drop_unspecified_subcategories
Applying strategy: link_biosphere_by_flow_uuid
Applying strategy: link_internal_technosphere_by_composite_code
Applying strategy: delete_exchanges_missing_activity
Applying strategy: delete_ghost_exchanges
Applying strategy: nuncertainty
Applied 13 strategies in 3.12 seconds
11301 datasets
521712 exchanges
780 unlinked exchanges
  Type biosphere: 12 unique unlinked exchanges


(11301, 521712, 780)

In [4]:
for x in ei.unlinked:
    print(x['name'])

Fluorene_temp
Fluoranthene_temp
Dibenz(a,h)anthracene_temp
Benzo(k)fluoranthene_temp
Benzo(ghi)perylene_temp
Benzo(b)fluoranthene_temp
Benzo(a)anthracene_temp
Acenaphthylene_temp
Chrysene_temp
Pyrene_temp
Phenanthrene_temp
Indeno(1,2,3-c,d)pyrene_temp


These temporary flows can be safely dropped.

In [5]:
ei.drop_unlinked(True)

Applying strategy: drop_unlinked
Applied 1 strategies in 0.16 seconds


In [6]:
ei.write_database()

Writing activities to SQLite3 database:
0%                          100%
[##############################] | ETA[sec]: 0.000 
Total time elapsed: 48.384 sec


Title: Writing activities to SQLite3 database:
  Started: 12/06/2016 10:53:25
  Finished: 12/06/2016 10:54:13
  Total time elapsed: 48.384 sec
  CPU %: 80.900000
  Memory %: 3.420854
Created database: 3.1 cutoff


Brightway2 SQLiteBackend: 3.1 cutoff

In [7]:
ei = SingleOutputEcospold2Importer(
    "/Users/cmutel/Documents/LCA Documents/Ecoinvent/3.2/cutoff/datasets", 
    "3.2 cutoff"
)
ei.apply_strategies()
ei.statistics()

Extracting XML data from 12916 datasets
Extracted 12916 datasets in 70.87 seconds
Applying strategy: normalize_units
Applying strategy: remove_zero_amount_coproducts
Applying strategy: remove_zero_amount_inputs_with_no_activity
Applying strategy: remove_unnamed_parameters
Applying strategy: es2_assign_only_product_with_amount_as_reference_product
Applying strategy: assign_single_product_as_activity
Applying strategy: create_composite_code
Applying strategy: drop_unspecified_subcategories
Applying strategy: link_biosphere_by_flow_uuid
Applying strategy: link_internal_technosphere_by_composite_code
Applying strategy: delete_exchanges_missing_activity
Applying strategy: delete_ghost_exchanges
Applying strategy: nuncertainty
Applied 13 strategies in 2.76 seconds
12916 datasets
459268 exchanges
0 unlinked exchanges
  


(12916, 459268, 0)

In [8]:
ei.write_database()

Writing activities to SQLite3 database:
0%                          100%
[##############################] | ETA[sec]: 0.000 
Total time elapsed: 38.414 sec


Title: Writing activities to SQLite3 database:
  Started: 12/06/2016 10:56:15
  Finished: 12/06/2016 10:56:54
  Total time elapsed: 38.414 sec
  CPU %: 88.000000
  Memory %: 5.748880
Created database: 3.2 cutoff


Brightway2 SQLiteBackend: 3.2 cutoff

In [9]:
ei = SingleOutputEcospold2Importer(
    "/Users/cmutel/Documents/LCA Documents/Ecoinvent/3.3/cutoff/datasets", 
    "3.3 cutoff"
)
ei.apply_strategies()
ei.statistics()

Extracting XML data from 13831 datasets
Extracted 13831 datasets in 121.34 seconds
Applying strategy: normalize_units
Applying strategy: remove_zero_amount_coproducts
Applying strategy: remove_zero_amount_inputs_with_no_activity
Applying strategy: remove_unnamed_parameters
Applying strategy: es2_assign_only_product_with_amount_as_reference_product
Applying strategy: assign_single_product_as_activity
Applying strategy: create_composite_code
Applying strategy: drop_unspecified_subcategories
Applying strategy: link_biosphere_by_flow_uuid
Applying strategy: link_internal_technosphere_by_composite_code
Applying strategy: delete_exchanges_missing_activity
Applying strategy: delete_ghost_exchanges
Applying strategy: nuncertainty
Applied 13 strategies in 3.17 seconds
13831 datasets
501148 exchanges
0 unlinked exchanges
  


(13831, 501148, 0)

In [10]:
ei.write_database()

Writing activities to SQLite3 database:
0%                          100%
[##############################] | ETA[sec]: 0.000 
Total time elapsed: 44.089 sec


Title: Writing activities to SQLite3 database:
  Started: 12/06/2016 10:59:53
  Finished: 12/06/2016 11:00:37
  Total time elapsed: 44.089 sec
  CPU %: 85.600000
  Memory %: 5.835319
Created database: 3.3 cutoff


Brightway2 SQLiteBackend: 3.3 cutoff

## 3.1 to 3.3

Which processes don't map exactly?

In [11]:
data_31 = {(ds['name'], ds['reference product'], ds['location']) for ds in Database("3.1 cutoff")}
data_33 = {(ds['name'], ds['reference product'], ds['location']) for ds in Database("3.3 cutoff")}

814 out of 11301 activities are not the same in 3.3.

In [12]:
len(data_31), len(data_31.difference(data_33))

(11301, 814)

Let's get a selection of the missing activities.

In [13]:
list(data_31.difference(data_33))[:100]

[('market for scrap steel', 'scrap steel', 'GLO'),
 ('electricity production, hard coal', 'electricity, high voltage', 'PL'),
 ('electricity production, photovoltaic, 3kWp flat-roof installation, multi-Si',
  'electricity, low voltage',
  'TW'),
 ('apple production', 'apple', 'GLO'),
 ('treatment of manure and biowaste by anaerobic digestion, from manure, liquid, cattle',
  'biogas',
  'RoW'),
 ('electricity, from municipal waste incineration to generic market for electricty, medium voltage',
  'electricity, medium voltage',
  'RoW'),
 ('field application of potassium nitrate',
  'potassium fertiliser, as K2O',
  'GLO'),
 ('electricity, from municipal waste incineration to generic market for electricty, medium voltage',
  'electricity, medium voltage',
  'NO'),
 ('market for sawnwood, lath, hardwood, raw, air dried',
  'sawnwood, lath, hardwood, raw, air dried',
  'GLO'),
 ('electricity, from municipal waste incineration to generic market for electricty, medium voltage',
  'electricity

## Official ecoinvent update list

Let's look at the official ecoinvent update Excel sheet. These are provided by ecoinvent in the "files" section (you have to be logged in; I have changed the filenames as well).

From these files, we need to get two things:

1. A list of activities which have had their names or reference products changed
2. A list of activities which have had major changes and need to be matched manually

In [29]:
import os
import pandas as pd
import math

### `Deleted or replaced` activities

In [23]:
overview = pd.read_excel("ecoinvent 3.1-3.2.xlsx", "overview")

In [24]:
overview['status'].unique()

array(['in both versions', 'new', 'deleted or replaced'], dtype=object)

In [25]:
overview

ISIC number                                         ISIC class  \
0            2011                     Manufacture of basic chemicals   
1            2011                     Manufacture of basic chemicals   
2              20     Manufacture of chemicals and chemical products   
3            2011                     Manufacture of basic chemicals   
4            2011                     Manufacture of basic chemicals   
5              20     Manufacture of chemicals and chemical products   
6            2011                     Manufacture of basic chemicals   
7            2011                     Manufacture of basic chemicals   
8            2011                     Manufacture of basic chemicals   
9            2011                     Manufacture of basic chemicals   
10           2011                     Manufacture of basic chemicals   
11           2011                     Manufacture of basic chemicals   
12           2011                     Manufacture of basic chemicals   
13           2011                     Manufacture of basic chemicals   
14             20     Manufacture of chemicals and chemical products   
15           2011                     Manufacture of basic chemicals   
16           2011                     Manufacture of basic chemicals   
17           2011                     Manufacture of basic chemicals   
18           2011                     Manufacture of basic chemicals   
19           2011                     Manufacture of basic chemicals   
20           2011                     Manufacture of basic chemicals   
21           2011                     Manufacture of basic chemicals   
22           2011                     Manufacture of basic chemicals   
23           2011                     Manufacture of basic chemicals   
24           2011                     Manufacture of basic chemicals   
25           2011                     Manufacture of basic chemicals   
26           2011                     Manufacture of basic chemicals   
27           2011                     Manufacture of basic chemicals   
28           2011                     Manufacture of basic chemicals   
29           2011                     Manufacture of basic chemicals   
...           ...                                                ...   
11844        0220                                            Logging   
11845        0220                                            Logging   
11846        0220                                            Logging   
11847        0220                                            Logging   
11848        1311         Preparation and spinning of textile fibres   
11849        1311         Preparation and spinning of textile fibres   
11850        1311         Preparation and spinning of textile fibres   
11851        1311         Preparation and spinning of textile fibres   
11852        1050                      Manufacture of dairy products   
11853        1050                      Manufacture of dairy products   
11854        2023  Manufacture of soap and detergents, cleaning a...   
11855        2023  Manufacture of soap and detergents, cleaning a...   
11856        2023  Manufacture of soap and detergents, cleaning a...   
11857        2023  Manufacture of soap and detergents, cleaning a...   
11858        2592         Treatment and coating of metals; machining   
11859        2592         Treatment and coating of metals; machining   
11860        2592         Treatment and coating of metals; machining   
11861        2592         Treatment and coating of metals; machining   
11862        2592         Treatment and coating of metals; machining   
11863        2592         Treatment and coating of metals; machining   
11864        2011                     Manufacture of basic chemicals   
11865        2011                     Manufacture of basic chemicals   
11866        2011                     Manufacture of basic chemicals   
11867        2011                     Manufact

Let's get a list of the activities which were renamed, but are otherwise not substantially changed. This will be from 3.1 to 3.2; we will need to combine this with the changes from 3.2 to 3.3.

In [26]:
renamed_mask = (overview['reference product renamed'] == 'yes') | (overview['activityName renamed'] == 'yes')
renamed_mask.sum()

172

Construct dictionary mapping names, reference products, and locations.

Need to be clever and handle `NaN` in some places where the reference product or activity name didn't change.

Also, change 'GLO' to 'RoW' where needed.

Will drop renamings (forcing a manual correction) if the 3.2 activity is not also in 3.3.

In [42]:
renamed_31_33 = {}

for _, row in overview[renamed_mask].iterrows():
    value = [row['activityName'], row['reference product'], row['geography']]

    if isinstance(row['activityName previous'], str):
        old_name = row['activityName previous']
    else:
        old_name = row['activityName']
    if isinstance(row['reference product previous'], str):
        old_rp = row['reference product previous']
    else:
        old_rp = row['reference product']
    key = [old_name, old_rp, row['geography']]
    
    if tuple(key) not in data_31 and key[2] == 'GLO':
        key[2] = 'RoW'
    if tuple(value) not in data_33 and value[2] == 'GLO':
        value[2] = 'RoW'
    
    if tuple(key) in data_31 and tuple(value) in data_33:
        renamed_31_33[tuple(key)] = tuple(value)

In [43]:
renamed_31_33

{('beam, hardwood, raw, air drying',
  'sawnwood, beam, hardwood, raw, air dried',
  'CH'): ('beam, hardwood, raw, air drying to u=20%',
  'sawnwood, beam, hardwood, raw, dried (u=20%)',
  'CH'),
 ('beam, hardwood, raw, air drying',
  'sawnwood, beam, hardwood, raw, air dried',
  'RoW'): ('beam, hardwood, raw, air drying to u=20%',
  'sawnwood, beam, hardwood, raw, dried (u=20%)',
  'RoW'),
 ('beam, hardwood, raw, kiln drying',
  'sawnwood, beam, hardwood, raw, kiln dried',
  'CH'): ('beam, hardwood, raw, kiln drying to u=10%',
  'sawnwood, beam, hardwood, raw, dried (u=10%)',
  'CH'),
 ('beam, hardwood, raw, kiln drying',
  'sawnwood, beam, hardwood, raw, kiln dried',
  'RoW'): ('beam, hardwood, raw, kiln drying to u=10%',
  'sawnwood, beam, hardwood, raw, dried (u=10%)',
  'RoW'),
 ('beam, softwood, raw, air drying',
  'sawnwood, beam, softwood, raw, air dried',
  'CH'): ('beam, softwood, raw, air drying to u=20%',
  'sawnwood, beam, softwood, raw, dried (u=20%)',
  'CH'),
 ('beam, s

## Export list of renamed activities, and missing activities

Easiest to make this one big Excel sheet.

In [45]:
import xlsxwriter

In [46]:
workbook = xlsxwriter.Workbook('ecoinvent-31-33-migration.xlsx')
worksheet = workbook.add_worksheet('matching')

labels = [
    "Old name",
    "Old RP",
    "Old location",
    "New name",
    "New RP",
    "New location",
    "Comment",
]

for index, label in enumerate(labels):
    worksheet.write(0, index, label)

for index, row in enumerate(sorted(data_31.difference(data_33))):
    for col in (0, 1, 2):
        worksheet.write(index + 1, col, row[col])
    if row in renamed_31_33:
        new = renamed_31_33[row]
        for col in (0, 1, 2):
            worksheet.write(index + 1, col + 3, new[col])

workbook.close()

## Export list of 3.3 activities for matching

In [47]:
workbook = xlsxwriter.Workbook('ecoinvent-33.xlsx')
worksheet = workbook.add_worksheet('data')

labels = [
    "Name",
    "RP",
    "Location",
]

for index, label in enumerate(labels):
    worksheet.write(0, index, label)

for index, row in enumerate(sorted(data_33)):
    for col in (0, 1, 2):
        worksheet.write(index + 1, col, row[col])

workbook.close()

# Now do the same thing for ecoinvent 3.2

In [50]:
overview = pd.read_excel("ecoinvent 3.2-3.3.xlsx", "overview")

In [51]:
data_32 = {(ds['name'], ds['reference product'], ds['location']) for ds in Database("3.2 cutoff")}

In [54]:
len(data_32), len(data_32.difference(data_33))

(12916, 319)

In [52]:
overview['status'].unique()

array(['new', 'present in both version without renaming',
       'present in both version with renaming', 'deleted'], dtype=object)

In [53]:
overview

ISIC number                                         ISIC class  \
0            2394            Manufacture of cement, lime and plaster   
1            0161             Support activities for crop production   
2              68                             Real estate activities   
3            0127                          Growing of beverage crops   
4            0810                  Quarrying of stone, sand and clay   
5            3822          Treatment and disposal of hazardous waste   
6            0127                          Growing of beverage crops   
7            0161             Support activities for crop production   
8            2394            Manufacture of cement, lime and plaster   
9            0161             Support activities for crop production   
10           3821      Treatment and disposal of non-hazardous waste   
11           0127                          Growing of beverage crops   
12           3830                                 Materials recovery   
13           3510  Electric power generation, transmission and di...   
14           3822          Treatment and disposal of hazardous waste   
15           0161             Support activities for crop production   
16           2220                   Manufacture of plastics products   
17           2394            Manufacture of cement, lime and plaster   
18           0125     Growing of other tree and bush fruits and nuts   
19           3822          Treatment and disposal of hazardous waste   
20           3700                                           Sewerage   
21             68                             Real estate activities   
22           2394            Manufacture of cement, lime and plaster   
23           3821      Treatment and disposal of non-hazardous waste   
24             68                             Real estate activities   
25             68                             Real estate activities   
26             68                             Real estate activities   
27             68                             Real estate activities   
28             68                             Real estate activities   
29             68                             Real estate activities   
...           ...                                                ...   
12409        0220                                            Logging   
12410        0220                                            Logging   
12411        0220                                            Logging   
12412        0220                                            Logging   
12413        1311         Preparation and spinning of textile fibres   
12414        1311         Preparation and spinning of textile fibres   
12415        1311         Preparation and spinning of textile fibres   
12416        1311         Preparation and spinning of textile fibres   
12417        1050                      Manufacture of dairy products   
12418        1050                      Manufacture of dairy products   
12419        2023  Manufacture of soap and detergents, cleaning a...   
12420        2023  Manufacture of soap and detergents, cleaning a...   
12421        2023  Manufacture of soap and detergents, cleaning a...   
12422        2023  Manufacture of soap and detergents, cleaning a...   
12423        2592         Treatment and coating of metals; machining   
12424        2592         Treatment and coating of metals; machining   
12425        2592         Treatment and coating of metals; machining   
12426        2592         Treatment and coating of metals; machining   
12427        2592         Treatment and coating of metals; machining   
12428        2592         Treatment and coating of metals; machining   
12429        2011                     Manufacture of basic chemicals   
12430        2011                     Manufacture of basic chemicals   
12431        2011                     Manufacture of basic chemicals   
12432        2011                     Manufact

In [55]:
renamed_mask = overview['status'] == 'present in both version with renaming'
renamed_mask.sum()

81

In [56]:
renamed_32_33 = {}

for _, row in overview[renamed_mask].iterrows():
    value = (row['activityName 3.3'], row['Product 3.3'], row['geography 3.3'])
    key = (row['activityName 3.2'], row['Product 3.2'], row['geography 3.2'])
    
    if key not in data_32 and key[2] == 'GLO':
        key = (key[0], key[1], 'RoW')
    if value not in data_33 and value[2] == 'GLO':
        value = (value[0], value[1], 'RoW')
    
    if key in data_32 and value in data_33:
        renamed_32_33[key] = value

In [58]:
renamed_32_33

{('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area 1'): ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area, Africa'),
 ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area 2, without Quebec'): ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area, North America, without Quebec'),
 ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area 3'): ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area, South America'),
 ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area 4&5 without China'): ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area, Asia, without China and GCC'),
 ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area 8'): ('aluminium production, primary, ingot',
  'aluminium, primary, ingot',
  'IAI Area, Gu

In [59]:
workbook = xlsxwriter.Workbook('ecoinvent-32-33-migration.xlsx')
worksheet = workbook.add_worksheet('matching')

labels = [
    "Old name",
    "Old RP",
    "Old location",
    "New name",
    "New RP",
    "New location",
    "Comment",
]

for index, label in enumerate(labels):
    worksheet.write(0, index, label)

for index, row in enumerate(sorted(data_32.difference(data_33))):
    for col in (0, 1, 2):
        worksheet.write(index + 1, col, row[col])
    if row in renamed_32_33:
        new = renamed_32_33[row]
        for col in (0, 1, 2):
            worksheet.write(index + 1, col + 3, new[col])

workbook.close()

# Validate manual lists

Need to make sure we can find each new linked activity.

In [60]:
import xlrd

In [67]:
ws = xlrd.open_workbook("/Users/cmutel/Downloads/ecoinvent-32-33-migration.xlsx").sheet_by_name("matching")
new_data = [
    tuple([ws.cell_value(row, col) for col in (3,4,5)]) 
    for row in range(1, ws.nrows) 
    if ws.cell_value(row, 3)
]

for key in new_data:
    if key not in data_33:
        print(key)

In [68]:
ws = xlrd.open_workbook("/Users/cmutel/Downloads/ecoinvent-31-33-migration.xlsx").sheet_by_name("matching")
new_data = [
    tuple([ws.cell_value(row, col) for col in (3,4,5)]) 
    for row in range(1, ws.nrows) 
    if ws.cell_value(row, 3)
]

for key in new_data:
    if key not in data_33:
        print(key)

How many of our 517 missing activities were `deleted or replaced`?

In [21]:
changed_31 = set(zip(overview['activityName'], overview['reference product'], overview['geography']))
517 - len(data_31.difference(data_32).difference(changed_31))

275

### Changed reference products

Find cases where the reference or byproduct name was changed. Byproducts need to be included because this list if for the undefined datasets - the byproducts can become reference product in new activities, depending on the system model.

In [25]:
changes = pd.read_excel(fp, "quantitative changes")

There are four types of changes listed here:

In [49]:
changes['change'].unique()

array(['updated', 'renaming only', 'added', 'deleted'], dtype=object)

We don't care about `added` activities, nor do we care about `updated` activities. Updated activities have changes in the amount, production volume, or price of an exchange, but we are trying to figure out how to update links from 3.1 to 3.2 - not update 3.1 itself. So we don't need to care about these changes.

Find changes where:

* The change was to a reference or byproduct
* The change was a renaming change

In [51]:
mask = (
    ((changes['group'] == 'ReferenceProduct') | (changes['group'] == 'Byproduct')) *
    (changes['change'] == 'renaming only')
)
mask.sum()

183

In [48]:
changes[mask]

ISIC number                                       ISIC class  \
480          1610                   Sawmilling and planing of wood   
481          1610                   Sawmilling and planing of wood   
483          1610                   Sawmilling and planing of wood   
633          1610                   Sawmilling and planing of wood   
1914         1610                   Sawmilling and planing of wood   
3111         1610                   Sawmilling and planing of wood   
3113         1610                   Sawmilling and planing of wood   
3115         1610                   Sawmilling and planing of wood   
3154         1610                   Sawmilling and planing of wood   
3155         1610                   Sawmilling and planing of wood   
3210         1610                   Sawmilling and planing of wood   
3240         1610                   Sawmilling and planing of wood   
4041         1610                   Sawmilling and planing of wood   
4282         1610                   Sawmilling and planing of wood   
4293         1610                   Sawmilling and planing of wood   
4815         4923                        Freight transport by road   
4898         1610                   Sawmilling and planing of wood   
4899         1610                   Sawmilling and planing of wood   
4900         1610                   Sawmilling and planing of wood   
5115         1610                   Sawmilling and planing of wood   
5116         1610                   Sawmilling and planing of wood   
5117         1610                   Sawmilling and planing of wood   
5662         1610                   Sawmilling and planing of wood   
5711         1610                   Sawmilling and planing of wood   
5713         1610                   Sawmilling and planing of wood   
5714         1610                   Sawmilling and planing of wood   
6127         1610                   Sawmilling and planing of wood   
6192         2392           Manufacture of clay building materials   
6193         2392           Manufacture of clay building materials   
6625         1610                   Sawmilling and planing of wood   
...           ...                                              ...   
31470        1610                   Sawmilling and planing of wood   
31516        1610                   Sawmilling and planing of wood   
31520        1610                   Sawmilling and planing of wood   
31707        1610                   Sawmilling and planing of wood   
31713        1610                   Sawmilling and planing of wood   
32427        1610                   Sawmilling and planing of wood   
32431        1610                   Sawmilling and planing of wood   
32901        1610                   Sawmilling and planing of wood   
32902        1610                   Sawmilling and planing of wood   
32991        1610                   Sawmilling and planing of wood   
33050        1610                   Sawmilling and planing of wood   
33402        1610                   Sawmilling and planing of wood   
33552        1610                   Sawmilling and planing of wood   
33590        1610                   Sawmilling and planing of wood   
33591        1610                   Sawmilling and planing of wood   
33592        1610                   Sawmilling and planing of wood   
33932        1610                   Sawmilling and planing of wood   
34006        4923                        Freight transport by road   
34367        1610                   Sawmilling and planing of wood   
34380        1610                   Sawmilling and planing of wood   
34893        1610                   Sawmilling and planing of wood   
35269        1610                   Sawmilling and planing of wood   
35270        1610                   Sawmilling and planing of wood   
35484        1610                   Sawmilling and planing of wood   
35598        2392           Manufacture of clay building materials   
35599        239

We need to check and make sure 1) all the old activity data is included in the 3.1 missing list, and 2) that the new activity data is included in the difference between 3.2 and 3.1.

In [58]:
import math

All old rows are in the set of missing 3.1 activities.

In [94]:
renamed_activities_old = set(zip(
        changes[mask]['activityName'], 
        changes[mask]['activityName previous'], 
        changes[mask]['exchange name previous'], 
        changes[mask]['geography'],
        changes[mask]['exchange name']
))

find_in = data_31.difference(data_32)
found_renamed_31 = {}

for row in renamed_activities_old:
    if not isinstance(row[1], str) and math.isnan(row[1]):
        old_name = row[0]
    else:
        old_name = row[1]
    new_name, old_rp, new_rp, location = row[0], row[2], row[4], row[3]

    if (old_name, old_rp, location) in find_in:
        found_renamed_31[(old_name, old_rp, location)] = (new_name, new_rp, location)
    elif location == 'GLO' and (old_name, old_rp, "RoW") in find_in:
        found_renamed_31[(old_name, old_rp, "RoW")] = (new_name, new_rp, "RoW")
    else:
        print(row)

In [95]:
len(found_renamed_31)

112

Check to make sure we can find the new activities in 3.2:

In [64]:
check_row = set(zip(
    changes[mask]['activityName'], 
    changes[mask]['exchange name'], 
    changes[mask]['geography'])
).difference(data_32)

assert {x[2] for x in check_row} == {'GLO'}
assert not {(x[0], x[1], 'RoW') for x in check_row}.difference(data_32)

### Current status

There are 11031 activities in ecoinvent 3.1 cutoff to upgrade to 3.2. Of these:

* 10784 are no problem - we can find the same activity name, reference product, and location
* 517 need more attention
* Of these 517, 112 were simple renames

Let's look at some of the remaining problematic activities:

In [96]:
list(data_31.difference(data_32).difference(set(found_renamed_31)))[:100]

[('electricity production, photovoltaic, 3kWp slanted-roof installation, CdTe, laminated, integrated',
  'electricity, low voltage',
  'MX'),
 ('electricity production, natural gas, at conventional power plant',
  'electricity, high voltage',
  'AU'),
 ('concrete production, for de-icing salt contact',
  'concrete, for de-icing salt contact',
  'CH'),
 ('wheat grain, feed production, Swiss integrated production',
  'wheat grain, feed, Swiss integrated production',
  'RoW'),
 ('electricity production, geothermal', 'electricity, high voltage', 'RU'),
 ('electricity production, photovoltaic, 3kWp slanted-roof installation, CdTe, laminated, integrated',
  'electricity, low voltage',
  'KR'),
 ('electricity production, natural gas, at conventional power plant',
  'electricity, high voltage',
  'MY'),
 ('planing, board, softwood, air dried',
  'shaving, softwood, measured as dry mass',
  'CH'),
 ('electricity production, photovoltaic, 3kWp slanted-roof installation, ribbon-Si, panel, mounted

## Write changes

In [92]:
import csv

In [97]:
with open('ecoinvent 31-32.csv', 'w') as csvfile:
    writer = csv.writer(csvfile)
    
    writer.writerow(["Old activity name", "Reference product", "Location",
                     "New activity name", "Reference product", "Location"])
    
    data = []
    
    for k, v in found_renamed_31.items():
        data.append(list(k) + list(v))
        
    for elem in data_31.difference(data_32).difference(set(found_renamed_31)):
        data.append(list(elem))
        
    data.sort()
    
    for line in data:
        writer.writerow(line)

In [98]:
with open('ecoinvent 32.csv', 'w') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(["Activity name", "Reference product", "Location"])
    for line in sorted(data_32):
        writer.writerow(line)